draconic/interpreter.py

```diff
@@ -645,12 +702,49 @@ def _assign_unpack(self, names, values):
                 raise DraconicValueError(
                     f"Cannot unpack non-iterable {type(values).__name__} object", names, self._expr
                 )
-            if not len(names.elts) == len(values):
-                raise DraconicValueError(
-                    f"Unequal unpack: {len(names.elts)} names, {len(values)} values", names, self._expr
-                )
-            for t, v in zip(names.elts, values):
-                self._assign_unpack(t, v)
+
+            stars = (i for i in names.elts if type(i) is ast.Starred)
+            starred = next(stars, None)
+
+            if starred is None:
+                if len(names.elts) > len(values):
+                    raise DraconicValueError(
+                        f"not enough values to unpack (expected {len(names.elts)}, got {len(values)})",
+                        values,
+                        self._expr,
+                    )
+                elif len(names.elts) < len(values):
+                    raise DraconicValueError(
+                        f"too many values to unpack (expected {len(names.elts)}, got {len(values)})",
+                        values,
+                        self._expr,
+                    )
+                for t, v in zip(names.elts, values):
+                    self._assign_unpack(t, v)
+            elif (extra := next(stars, None)) is not None:
+                raise DraconicSyntaxError.from_node(extra, "multiple starred expressions in assignment", self._expr)
+            else:
+                if len(values) < (len(names.elts) - 1):
+                    raise DraconicValueError(
+                        f"not enough values to unpack (expected at least {len(names.elts) - 1}, got {len(values)})",
+                        values,
+                        self._expr,
+                    )
+                for t, v in zip(names.elts, values):
+                    if t is starred:
+                        break
+                    values.remove(v)
+                    self._assign_unpack(t, v)
+
+                for t, v in zip(reversed(names.elts), reversed(values)):
+                    if t is starred:
+                        break
+                    values.remove(v)
+                    self._assign_unpack(t, v)
+                self._assign_unpack(starred, values)
+
+    def _assign_starred(self, name, value):
+        self._assign_name(name.value, value)
```

Comment:
What is this doing?

Reply:
It seems it was a mistake, it should be throwing a syntax error

Reply:
That method is called whenever anything is assigned to any starred thing, added to the assignment targets dict here: https://github.com/avrae/draconic/blob/eb47655ac2a36caea4f64f4d162f911496d10cd6/draconic/interpreter.py#L394

I don't know if there is a better way to do this, since assignment is handled through those nodes only.
This does lead to the consequence where
```py
*a = [1, 2, 3]
a == [1, 2, 3]
```
as opposed to a SyntaxError like normal Python.

Reply:
Interestingly, though it is valid syntax (`*a = (1, 2)` parses into this[^1]), the Python behaviour does seem to be throwing a SyntaxError.

[^1]: 
    ```
    Module(
       body=[
          Assign(
             targets=[
                Starred(
                   value=Name(id='a', ctx=Store()),
                   ctx=Store())],
             value=Tuple(
                elts=[
                   Constant(value=1),
                   Constant(value=2)],
                ctx=Load()))],
       type_ignores=[])
     ```

Reply:
> That method is called whenever anything is assigned to any starred thing, added to the assignment targets dict here:
> 
> https://github.com/avrae/draconic/blob/eb47655ac2a36caea4f64f4d162f911496d10cd6/draconic/interpreter.py#L394
> 
> I don't know if there is a better way to do this, since assignment is handled through those nodes only. This does lead to the consequence where
> 
> ```python
> *a = [1, 2, 3]
> a == [1, 2, 3]
> ```
> 
> as opposed to a SyntaxError like normal Python.

Isn't the example in the comment assigning to a tuple (and therefore would go through `_assign_unpack`) rather than Starred?
```
Module(
   body=[
      Assign(
         targets=[
            Tuple(
               elts=[
                  Name(id='a', ctx=Store()),
                  Starred(
                     value=Name(id='b', ctx=Store()),
                     ctx=Store())],
               ctx=Store())],
         value=Tuple(
            elts=[
               Constant(value=1),
               Constant(value=2),
               Constant(value=3)],
            ctx=Load()))],
   type_ignores=[])
```

Reply:
Hmm, this is more problematic than it was at first glace

Reply:
In python `*a = [1, 2, 3]` is technically invalid syntax, however the AST makes you handle it yourself, which is problematic, because we also need this function to assign properly when the syntax is correct, and we can't think of a simple way to check if the starred node is inside of a list/tuple or not.

Reply:
At the very minimum I know that commenting out the `ast.Starred: self._assign_starred,` breaks everything because `FeatureNotAvailable`